### PR TITLE
[oneDNN] Enable BF16 for AvgPool3D, MaxPool3D, AvgPool3DGrad, MaxPool3DGrad

### PR DESCRIPTION
**Pull request by gaurides**
Enable BF16 Eigen kernels for AvgPool3D, MaxPool3D, AvgPool3DGrad, MaxPool3DGrad ops

### CODE REVIEW

**Comment by gaurides** [comment]
This PR basically fixes the 3d pooling kernels by removing some "float" type hardcoding and uses the template types. Also, fixes the test to test for BF16 datatype.

Changes in couple other files are minor - coding style and comments.
This PR is on same lines as PR 56929. Hope this helps to review the PR.

**Comment by nalinik** [comment]
Is there progress on merging this into the main branch? What's the milestone?

**Comment by gaurides** [comment]
Hi @penpornk : Can you please let me know what is the error here? I see ARM CI was successful but ARM CI Extended has issue. Do we need a guard here too?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 1]
```diff
@@ -68,26 +76,27 @@ def _VerifyOneTest(self, pool_func, input_sizes, window, strides, padding,
...
-        t = test_util.NCHWToNHWC(t)
-      vals = self.evaluate(t)
-    # Verifies values.
-    actual = vals.flatten()
-    self.assertAllClose(expected, actual)
+    for dtype in self._DtypesToTest(use_gpu):
+      with self.cached_session(use_gpu=use_gpu):
+        t = constant_op.constant(x, shape=input_sizes, dtype=dtype)
+        ksize = [1] + list(window) + [1]
+        strides_a = [1] + list(strides) + [1]
```
[the remark this answers is not in the document]
Comment (gaurides) [reply tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py:83]: Thanks for the review @penpornk .
Now that I added a for loop to handle multiple data types, the output variable strides becomes a local variable, else the update strides will be used for the next datatype in the for loop.
It's just like the "ksize" variable derived from "window".

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 2]
```diff
@@ -270,59 +281,78 @@ def _ConstructAndTestGradientForConfig(self,
...
-
-      ksize = [1, window[0], window[1], window[2], 1]
-      strides = [1, strides[0], strides[1], strides[2], 1]
-      t = input_tensor
-
-      if data_format == "NCDHW":
-        ksize = test_util.NHWCToNCHW(ksize)
-        strides = test_util.NHWCToNCHW(strides)
-        t = test_util.NHWCToNCHW(t)
-        output_sizes = test_util.NHWCToNCHW(output_sizes)
```
[the remark this answers is not in the document]
Comment (gaurides) [reply tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py:296]: yes, it was intentional and then i was converting the format back later. But it's not required. So, i will add this line back and remove the format conversion later.

**Comment by gaurides** [comment]
Hi @penpornk - i addressed your review comments. 
I accidentally removed @qqfish from the reviewers. Can you please add him back? Thanks.

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 3]
```diff
@@ -270,59 +281,78 @@ def _ConstructAndTestGradientForConfig(self,
...
+
+        t_p = pool_func(
+            t,
+            ksize=ksize,
+            strides=strides_a,
+            padding=padding,
+            data_format=data_format,
+            name=func_name)
+        if data_format == "NCDHW":
+          t_p = test_util.NCHWToNHWC(t_p)
```
[the remark this answers is not in the document]
Comment (gaurides) [reply tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py]: removed this conversion, and added the output_sizes conversion back.

**Comment by gaurides** [comment]
Resolved the merge conflict.

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 4]
```diff
@@ -104,6 +112,9 @@ def _VerifyValues(self, pool_func, input_sizes, window, strides,
       if use_gpu and test_util.is_gpu_available(cuda_only=True):
         self._VerifyOneTest(pool_func, input_sizes, window, strides, padding,
                             data_format, dtypes.bfloat16, expected, use_gpu)
+      elif not use_gpu:
+        self._VerifyOneTest(pool_func, input_sizes, window, strides, padding,
+                            data_format, dtypes.bfloat16, expected, use_gpu)
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py:117]: The only case where bf16 isn't run is when `use_gpu` is `true` and `is_gpu_available` is `false`. I think we can just merge the two `if`s.
```suggestion
      # Don't test bfloat16 on GPU if there is no GPU available.
      if (not use_gpu) or test_util.is_gpu_available(cuda_only=True):
        self._VerifyOneTest(pool_func, input_sizes, window, strides, padding,
                            data_format, dtypes.bfloat16, expected, use_gpu)
```

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 5]
```diff
@@ -138,19 +149,21 @@ def testAvgPool3dSamePaddingDifferentStrides(self):
...
-        t = gen_nn_ops.AvgPool3DGrad(
-            orig_input_shape=orig_input_shape,
-            grad=grad,
-            ksize=[1, 1, 1, 1, 1],
-            strides=[1, 1, 1, 1, 1],
-            padding="SAME",
-            data_format="NDHWC")
-        self.evaluate(t)
+      use_gpu = test.is_gpu_available(cuda_only=True)
+      for dtype in self._DtypesToTest(use_gpu):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py:153]: Why do we exclude `bfloat16` on GPU? I think we should test both `float32` and `bfloat16` on both CPU and GPU. (If so, we can probably just use a list here. We don't need to create `_DtypesToTest` if it's only used once.)

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 6]
```diff
@@ -48,6 +48,14 @@ def GetTestConfigs():
 @test_util.with_eager_op_as_function
 class PoolingTest(test.TestCase):
 
+  def _DtypesToTest(self, use_gpu):
+    if use_gpu:
+      return [dtypes.float32]
+    else:
+      # It is important that float32 comes before float16 here, as we will be
+      # using its gradients as reference for bf16 gradients.
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py:56]: This PR doesn't seem to use fp32 values to check bf16 values anymore. Should we remove this comment / order enforcement?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes and sorry for the delay!

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py (hunk 5 above)
Comment (gaurides) [reply tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py]: Thanks for the review @penpornk . You have valid suggestions, i addressed them all and pushed another commit.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the quick changes! One minor nit please.

File: tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py [hunk 7]
```diff
@@ -138,19 +139,21 @@ def testAvgPool3dSamePaddingDifferentStrides(self):
...
-            1, shape=[1, 3, 2, 4, 2], dtype=dtypes.float32)
-        t = gen_nn_ops.AvgPool3DGrad(
-            orig_input_shape=orig_input_shape,
-            grad=grad,
-            ksize=[1, 1, 1, 1, 1],
-            strides=[1, 1, 1, 1, 1],
-            padding="SAME",
-            data_format="NDHWC")
-        self.evaluate(t)
+      use_gpu = test.is_gpu_available(cuda_only=True)
```
[the remark this answers is not in the document]
Comment (gaurides) [reply tensorflow/python/kernel_tests/nn_ops/pooling_ops_3d_test.py]: yes, missed that. Thanks, i removed it now

**Review by penpornk — APPROVED** [review]
Thank you!

**Comment by penpornk** [comment]
[ARM CI failures](https://github.com/tensorflow/tensorflow/actions/runs/3883911445/jobs/6626246458) seem to be already failing tests (7 tests; no new failures).

**Pull request merged**